source/dofs/dof_accessor_set.cc

```diff
@@ -41,91 +41,235 @@
+DeclException2(ExcNonMatchingElementsSetDofValuesByInterpolation,
+               Number,
+               Number,
+               << "Called set_dof_values_by_interpolation(), but"
+               << " the element to be set, value " << std::setprecision(16)
+               << arg1 << ", does not match with the value "
+               << std::setprecision(16) << arg2 << " already set before.");
+
+namespace internal
+{
+  template <typename Number>
+  typename std::enable_if<
+    !std::is_unsigned<Number>::value,
+    typename numbers::NumberTraits<Number>::real_type>::type
+  get_abs(const Number a)
+  {
+    return std::abs(a);
+  }
+
+  template <typename Number>
+  typename std::enable_if<std::is_unsigned<Number>::value, Number>::type
+  get_abs(const Number a)
+  {
+    return a;
+  }
+
+  template <typename VectorType>
+  constexpr bool is_dealii_vector =
+    std::is_same<VectorType,
+                 dealii::Vector<typename VectorType::value_type>>::value ||
+    std::is_same<VectorType,
+                 dealii::BlockVector<typename VectorType::value_type>>::value ||
+    std::is_same<
+      VectorType,
+      dealii::LinearAlgebra::Vector<typename VectorType::value_type>>::value ||
+    std::is_same<VectorType,
+                 dealii::LinearAlgebra::distributed::Vector<
+                   typename VectorType::value_type>>::value ||
+    std::is_same<VectorType,
+                 dealii::LinearAlgebra::distributed::BlockVector<
+                   typename VectorType::value_type>>::value;
+
+  template <typename T>
+  using set_ghost_state_t =
+    decltype(std::declval<T const>().set_ghost_state(std::declval<bool>()));
+
+  template <typename T>
+  constexpr bool has_set_ghost_state =
+    is_supported_operation<set_ghost_state_t, T>;
+
+  template <typename VectorType,
+            typename std::enable_if<has_set_ghost_state<VectorType>,
+                                    VectorType>::type * = nullptr>
+  void
+  set_ghost_state(VectorType &vector, const bool ghosted)
+  {
+    vector.set_ghost_state(ghosted);
+  }
+
+  template <typename VectorType,
+            typename std::enable_if<!has_set_ghost_state<VectorType>,
+                                    VectorType>::type * = nullptr>
+  void
+  set_ghost_state(VectorType &, const bool)
+  {
+    // serial vector: nothing to do
+  }
+
+  template <int  dim,
+            int  spacedim,
+            bool lda,
+            class OutputVector,
+            typename number>
+  void
+  set_dof_values_with_check(const DoFCellAccessor<dim, spacedim, lda> &cell,
+                            const Vector<number> &local_values,
+                            OutputVector &        values,
+                            const bool            perform_check)
+  {
+    (void)perform_check;
+
+#ifdef DEBUG
+    if (perform_check && is_dealii_vector<OutputVector>)
+      {
+        set_ghost_state(values, true);
+
+        Vector<number> tmp(cell.get_fe().n_dofs_per_cell());
```

Comment:
```suggestion
        Vector<number> dof_values(cell.get_fe().n_dofs_per_cell());
```
or `cell_dof_values`. I know `tmp` was in the old code, but it would be nice to give it a better name here :)